sgkit/stats/pca.py

```diff
@@ -0,0 +1,341 @@
+        )
+
+    numblocks = da.asarray(_allele_counts(ds, variable, check_missing=False)).numblocks
+    if all(s > 1 for s in numblocks) and algorithm != "randomized":
+        raise ValueError(
+            "PCA can only be performed on arrays chunked in 2 dimensions if algorithm='randomized'. "
+            "Consider using this algorithm instead or rechunking the alternate allele counts array "
+            "(e.g. ds.call_alternate_allele_count.chunk((None, -1)))."
+        )
+
+    return Pipeline(
+        [
+            ("scaler", PattersonScaler(ploidy=ploidy)),
+            (
+                "pca",
+                TruncatedSVD(
+                    n_components=n_components,
+                    algorithm=algorithm,
+                    n_iter=n_iter,
+                    random_state=random_state,
+                    compute=False,
+                ),
+            ),
+        ]
+    )
+
+
+def pca_fit(
+    ds: Dataset,
+    est: BaseEstimator,
+    *,
+    variable: str = "call_alternate_allele_count",
+    check_missing: bool = True,
+) -> BaseEstimator:
+    """ Fit PCA estimator """
+    AC = _allele_counts(ds, variable, check_missing=check_missing)
+    return est.fit(da.asarray(AC).T)
+
+
+def pca_transform(
+    ds: Dataset,
+    est: BaseEstimator,
+    *,
+    variable: str = "call_alternate_allele_count",
+    check_missing: bool = True,
+    merge: bool = True,
+) -> Dataset:
+    """ Apply PCA estimator to new data """
+    AC = _allele_counts(ds, variable, check_missing=check_missing)
+    projection = est.transform(da.asarray(AC).T)
+    new_ds = Dataset({"sample_pca_projection": (("samples", "components"), projection)})
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def _get(est: BaseEstimator, attr: str, fn: Any = lambda v: v) -> Optional[ArrayLike]:
+    try:
+        if hasattr(est, "named_steps"):
+            est = est["pca"]
+        return fn(getattr(est, attr))
+    except Exception:
+        return None
+
+
+def pca_stats(ds: Dataset, est: BaseEstimator, *, merge: bool = True) -> Dataset:
+    """ Extract attributes from PCA estimator """
+    new_ds = {
+        "sample_pca_component": (
+            ("variants", "components"),
+            _get(est, "components_", fn=lambda v: v.T),
+        ),
+        "sample_pca_explained_variance": (
+            "components",
+            _get(est, "explained_variance_"),
+        ),
+        "sample_pca_explained_variance_ratio": (
+            "components",
+            _get(est, "explained_variance_ratio_"),
+        ),
+    }
+    new_ds = Dataset({k: v for k, v in new_ds.items() if v[1] is not None})
+    if "sample_pca_component" in new_ds and "sample_pca_explained_variance" in new_ds:
+        new_ds["sample_pca_loading"] = new_ds["sample_pca_component"] * np.sqrt(
+            new_ds["sample_pca_explained_variance"]
+        )
+    return conditional_merge_datasets(ds, new_ds, merge)
+
+
+def pca(
```

Comment:
Should this function use `sgkit.variables`, which would mean:
 * using variable references instead of strings
 * validate input
 * validate output
 * update the docs

Reply:
Added the variables (https://github.com/pystatgen/sgkit/pull/262/commits/05fa77df1a6c8f1fe110631dfd8672e526ad170f#diff-cd77642bae81c13ab776800efe4ba498900a284bc2e4741d943b727f60b8b7e1R250-R284) and updated references (https://github.com/pystatgen/sgkit/pull/262/commits/05fa77df1a6c8f1fe110631dfd8672e526ad170f#diff-e66844198337edbcfa8f27fbcbcc58ef011272669fe9295119b0847c1b7cdf69R134).   Squashed one too many commits before rebasing on the current master so I lost a clean delta, but that was basically all I changed.